documentation/man/man1/tri2fe.1
'\" t
.TH TRI2FE 1 "August 26, 2013" "Utrecht University"
.UC 4
.SH NAME
tri2fe -n TriangleNodeFile -e TriangleElmFile
.SH DESCRIPTION
Converts output from \fItriangle\fP to input for \fIGTECTON\fP. Nodal point coordinates are output to file "tecin.dat.nps" and element connectivity to file "tecin.dat.elm". 
These filenames cannot be changed.
.SH SEE ALSO
triangle(1), pln(1), pln(7), opn(1), opn(7), plnplt(1), opnplt(1)
.SH "BUGS OR POSSIBLE FLAWS"
Report bugs to Rob Govers (email: r.govers@uu.nl).


Changelog
Version     Date                 FEM/PLT   Change

1.4         1989                   ALL     Original version
1.51        September 1991         FEM     Updated plastic code, residual loads.
2.00        October/November 1991  FEM     Added thermal code.

2.01        March 1992             FEM     Debugged thermo-mechanical version
                                           Diffusion creep code.
2.02        August 1992            FEM     Linked nodes/no-tilt boundaries
2.03        October 7 1992         FEM     Consistent plane strain thermal expansion loads.
2.04        November 2 1992        FEM     Solution to diffusion eqn for surface uplifts
                                           and sedimentary transport.
2.05        November 6 1992        FEM     ITMODE=2 (initial temperatures)

2.06        June 7 1993            FEM     Array bounds checking
                                   FEM     Gravity using Talwani's method
2.07        July 1 1993            FEM     SURF arrays always read in. NSED=switch.
2.08        July 13 1993           FEM     Cohesion calculation based upon actual surface
                                           height
2.09        August 20 1993         FEM     Cohesion factor definition changed to include
                                           density
2.10        November 1993          FEM     Slippery nodes update (dimensioning)

2.10        March 1994             FEM     ICVIS switch added
2.11        June 1994              FEM     Thermal consequences of faults
2.12        November 1994          FEM     Sparse solvers (Petsc)

2.12        October 1995           FEM     Bug in ADDSN for triangles

2.13        May 1997               FEM     Incompatible modes and fault intersections

98.1        April 28 1998          ALL     Various modifications
                                   FEM     Out-of-plane library

99.1i       May 1999               FEM     Iterative solvers PetSc

2000        August 2000            FEM     Spherical shell elements

2001        July 2001              FEM     Consistent stress invariants calculations
                                   FEM     Restart facility

2003        June 2003              FEM     Tetrahedral elements

2005        November 2005          FEM     Petsc-2.2.1

2006        June 2006              PLT     GMT Support    

2008        January 2008           FEM     Incompressible plastic flow

2009        June 2009              FEM     Multiple output files (one per time step)

2010        April 2010             FEM     Incorporation Petsc-2.3.2 or higher
                                   FEM     Enabling of parallel computing
                                   FEM     Enhanced sparsity pattern usage for parallel computing

2013        May 2013               FEM     Multinode code runs
                                   ALL     Filepaths no longer restricted to 80 chars. In stead to 255

2014        February 2014          ALL     Upgraded to Fortran 2008 standard
                                   FEM     PETSc 3.4.2

2015        March 2015             ALL     Added axisymmetric solver and plotting

2016        September 2016         ALL     Various bugs

2017.3      March 2017             ALL     Slippery nodes working on multiptocessor runs.
                                           Most of the implicit variables removed.
                                           Part of the static memory allocation removed.

2017.9      September 2017         ALL     Polysort has maximum polygon size increased from 10k to 10M
                                           bugfix: plnplt 'slip' now gives good coordinates
                                           bugfix: plnplt 'nbc' command now returns values.
                                           bugfix: FEDSK files now contain correct velocities in ascii,
                                           More implicitness removed, implicitly fixing bugs.
                                           bugfix: Plt3d now correctly computes centers of elements.
                                           bugfix: Plnplt now correctly writes strain rate files (rate command)
                                           bugfix: ...and elemenental boundary condition file (ebc command)
                                           plnplt now accepts blank lines in scripts
                                           plnplt and plt3d help now take '? set' beside '? se'
                                           elmside now gives an error if node or elem file is empty
                                           Now use system soelim. Minimum version 1.21, 
                                                 in stead of own implementation
                                           bugfix: Elmside now scans element array with element index,
                                                 and nodal point arrays with nodal point index, instead
                                                 of both with the element index.
                                           The main storage is now dynamically allocated. This limits
                                                 the data size to the machine memory size, in stead of to
                                                 the 2GB maximum. These allocations can be set in module 
                                                 TempSrc/AAAB.F
                                           bugfix: The content of environment variable MYTMP is now used as 
                                                 temporary directory. Somehow this feature broke down in the
                                                 past.
                                           More implicit variables removed.
                                           Plt3d has the 'edi' command fixed for total, x, y and z 
                                                 displacements. Slippery and split displacements remain

src/01_meshing/2D/triangle-helpers/tri2fe.f
program main
! This program aims to convert a grid as described by triangle into a 
! grid as described by gtecton. For this it read the the domain.[node/ele/poly]
! files and it creates tecin.dat.[nps/ele]
!

implicit none

character(len=80) Nfile,Efile
logical :: writeSerialMode

!double precision :: attr(3)
integer           :: attr(3)
integer           :: m(3)
integer, external :: NINT,iflu,nextlu
double precision  :: x, y
integer           :: i, iattr, ibound, idim
integer           :: im, imark, ios, ip, np, ne, k
integer           :: luo, lui, lue
integer           :: npoints

writeSerialMode = .false.

call shell (Nfile,Efile,writeSerialMode)

lue = iflu('stderr')
lui = nextlu(0)

call openf(lui,Nfile,'old')
read (lui,*,err=990,end=1000,iostat=ios) np,idim,iattr,ibound

if (iattr.gt.0) then
    write(lue,1)
endif

1    format('node attributes not carried along')

if (ibound.eq.0) then
    write(lue,2)
2        format('external boundary nodes should be labeled <> 0')
    call exitp(1)
endif

!********* Do the nodes

luo = nextlu(lui+1)
call openf(luo,'tecin.dat.nps','unknown')

do k=1,np
    read (lui,*,err=990,end=1000,iostat=ios) ip,x,y,imark

    if (writeSerialMode) then
        write (luo,33) k,imark,x,y
    else
        write (luo,3) k,imark,x,y
    endif
enddo

!3      format (I12,I12,2E14.6)
3      format (I12,I12,2E25.17)  ! to match the output precision of triangle.
!    3  format(I12,I12,2E14)
33     format (I5,I5,2(1PG14.6)) ! serial old sk00l

write(luo,'(a)') 'end nps'
call closef(luo)
call closef(lui)
write(*,*) 'NUMNP = ',np

!******** Do the elements

call openf(lui,Efile,'old')
read (lui,*,err=990,end=1000,iostat=ios) ne,npoints,iattr

if (npoints.ne.3) then
    write(lue,4)
4        format('3 points per triangle supported')
    call exitp(1)
endif

if (iattr.gt.3) then
    write(lue,5)
5        format('a maximum of 3 attributes per triangle is supported')
endif
  
call openf(luo,'tecin.dat.elm','unknown')

attr(1) = 1
attr(2) = 0
attr(3) = 0

do k=1,ne

   if (iattr.gt.0) then
       read (lui,*,err=990,end=1000,iostat=ios) im,(m(i),i=1,3), (attr(i),i=1,iattr)
   else
       read (lui,*,err=990,end=1000,iostat=ios) im,(m(i),i=1,3)
   endif

   do i=1,3
         m(i) = m(i) + 1
         if (m(i).gt.np) then
            write(lue,7)
7           format('node numbering not zero-based')
            call exitp(1)
         endif
   enddo

   if (writeSerialMode) then 
       write (luo,66) k, attr(1), (m(i),i= 1,3),m(3)
   else
       write (luo,6) k, attr(1), (m(i),i= 1,3),m(3)
   endif
end do

6       format (6I12)
66       format (6I5)


write(luo,'(a)') 'end elm'

call closef(luo)
call closef(lui)

write(*,*) 'NUMEL = ',ne

call exitp(0)

990    write(lue,*) 'read error ',ios
call exitp(1)
1000    write(lue,*) 'EOF error ',ios
call exitp(1)
end
!-----------------------------------------------------------------------
subroutine shell (Nfile,Efile,writeSerialMode)
!
implicit none
!-pass
character*(*) Efile,Nfile
logical :: writeSerialMode
!-locl
character(len=80) usage,arg
integer iflu,stderr,narg,lnblk
external iflu,lnblk
!
!        call HowToUse ('-n TriangleNodeFile -e TriangleElmFile',usage)
stderr = iflu('stderr')
!
Nfile = ''
Efile = ''

narg = command_argument_count()

if (narg.ne.4 .and. narg.ne.5) then
write(stderr,'(80a)') usage(1:lnblk(usage))
call exitp(1)
endif

call get_command_argument(1,arg)

if (arg.eq.'-n') then
call get_command_argument(2,Nfile)
else if (arg.eq.'-e') then
call get_command_argument(2,Efile)
else
write(stderr,'(80a)') usage(1:lnblk(usage))
call exitp(1)
endif

call get_command_argument(3,arg)

if (arg.eq.'-n') then
call get_command_argument(4,Nfile)
else if (arg.eq.'-e') then
call get_command_argument(4,Efile)
else
write(stderr,'(80a)') usage(1:lnblk(usage))
call exitp(1)
endif

if (narg.eq.5) then
call get_command_argument(5,arg)
if (arg.eq.'serial') then
    writeSerialMode = .true.
endif
endif

if (lnblk(Nfile).eq.0 .or. lnblk(Efile).eq.0) then
write(stderr,'(80a)') usage(1:lnblk(usage))
call exitp(1)
endif

return
end
